modin/engines/modin_dataframe.py

```diff
@@ -0,0 +1,477 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import List, Hashable, Optional, Callable, Union, Dict
+
+
+class ModinDataframe(ABC):
```

Comment:
I think we should change `ModinDataframe` to `ModinDataFrame`.

Reply:
By the way, there was a discussion on removing part `Data` from `Modin DataFrame` some time ago. What way do we go?
@devin-petersohn , @vnlitvinov

Reply:
`DataFrame` is camel case for "data frame", but I think it should be one word. There are historical reasons for this.

`ModinFrame` is more concise, but I wouldn't call a Modin database `ModinBase`. Interested in other opinions too on this.

Reply:
I think calling something a `DataFrame` (or `Dataframe`) when we provide `modin.pandas.DataFrame` as `pandas.DataFrame` replacement looks ambiguous.

For that matter I think I would prefer seeing `BaseFrame` here, or even something without "Frame" word in it to remove even smallest possible confusion here, like `ModinBaseBody` maybe?

Reply:
> I think calling something a `DataFrame` (or `Dataframe`) when we provide `modin.pandas.DataFrame` as `pandas.DataFrame` replacement looks ambiguous.

Probably, yes. It looks ambiguous a bit.

> For that matter I think I would prefer seeing `BaseFrame` here, or even something without "Frame" word in it to remove even smallest possible confusion here, like `ModinBaseBody` maybe?

I wouldn't tie `Modin` part to any name of a class/method. I like name `BaseFrame`. Also, if we decide to support another distributed object at one day, it would be good to name base class like `BaseTensor`.

Reply:
>I wouldn't tie Modin part to any name of a class/method. I like name BaseFrame.

That makes sense. `BaseDataframe` or `CoreDataframe`. I don't think shorthanding `Dataframe` to `Frame` is worth the extra characters saved. I'm necessary agree that there's ambiguity here with the API layer here because users will not see this for the most part. Developers will understand the abstractions (provided we document them well enough 😄).

Reply:
I think I see your idea about saving `DataFrame` instead of `Frame`. Let's keep `DataFrame` then :)